### ISSUE TITLE
[feature] ignore URL encode

### ISSUE THREAD

**Issue by Freed-Wu** [codespell-project/codespell#2243]
The link of markdown should be a URL encode, such as a first header of markdown is

```
# 区别
```

mdtoc will generate a table of contents like

```
- [区别](#%E5%8C%BA%E5%88%AB)
```

However,

```
codespell.....................................................................Failed
- hook id: codespell
- exit code: 65

README.md:16: BA ==> BY, BE
```

Can codespell ignore that URL encoded characters?

**Comment by peternewman** [comment]
There's an URI regex:
https://github.com/codespell-project/codespell/blob/2674a1ddd8356946f1fa400a7baa57289cabaa64/codespell_lib/_codespell.py#L30-L34
And option to set your own:
https://github.com/codespell-project/codespell/blob/2674a1ddd8356946f1fa400a7baa57289cabaa64/codespell_lib/_codespell.py#L325-L328

But the default one doesn't match hash/anchors only.

There's also an ignore regex option so you could strip them out entirely (so they're not checked at all).

**Comment by Freed-Wu** [comment]
Thanks for your answer. Use `--ignore-regex` can ignore some strings like `#%E5%8C%BA%E5%88%AB`. However, can these strings are ignored by default by codespell? I believe most people all need to ignore these weird strings.

**Comment by yarikoptic** [comment]
and may be ignore also `@` (and other extra symbols?) which could be found used in URLs without encoding?

```
❯ codespell spellme.md
spellme.md:1: thi ==> the, this
❯ cat spellme.md
RLE implemented by [rle-pack](https://www.npmjs.com/package/@thi.ng/rle-pack) library.
```

**Comment by DimitriPapadopoulos** [comment]
@yarikoptic With `--uri-ignore-words-list` you can provide a list of words to ignore in URIs only.

Perhaps there is a misunderstanding here: codespell does not skip URIs by default, it will just identify them URIs as URIs. It also accepts an additional list of words to ignore specifically in URIs. It looks like you may provide `*` as an argument to `--uri-ignore-words-list`, in which case codespell will in effect skip URIs:
https://github.com/codespell-project/codespell/blob/a70c628036d2cf52f31266e5d2a7652e5c7e9ba4/codespell_lib/_codespell.py#L848-L863

In https://github.com/yarikoptic/label-studio/commit/8a07f4f6702213c3e3ae90d815bdb0a9927a0018, changing `@thi` to `%40thi` resolves the issue, **not** because `@` was not identified as part of a URI, but because the new resulting word `40thi` is not part of our dictionary of spelling errors. The same would happen outside URIs:
```
$ cat > foo.bar
@thi
$ codespell foo.bar 
foo.bar:1: thi ==> the, this
$ 
$ cat > foo.bar
%40thi
$ codespell foo.bar 
$ 

```

Does the `--uri-ignore-words-list=*` trick work for you? I have to agree, this is not well documented – just discovered this.

**Comment by DimitriPapadopoulos** [comment]
> However, can these strings are ignored by default by codespell? I believe most people all need to ignore these weird strings.

@Freed-Wu How do you formally define “_these weird strings_” you want to ignore? I mean, I understand the kind of strings you're thinking of, but we need actual code to identify “_these weird strings_” and avoid false positives, without false negatives in other situations.

**Comment by yarikoptic** [comment]
> Does the `--uri-ignore-words-list=*` trick work for you?

well -- it does "work" in that it behaves like you suggest!

```
❯ cat spellme.md
RLE implemented by [rle-pack](https://www.npmjs.com/package/@thi.ng/rle-pack) library.
❯ codespell spellme.md
spellme.md:1: thi ==> the, this
❯ codespell --uri-ignore-words-list=* spellme.md
 ```

but I indeed then misunderstood the URI aspect -- didn't realize codespell "spells" URIs as well by default (not sure how it is valuable generally if we do not have control over external URIs -- we can't just "fix" typos in them, unless internal within project - thus derived from filenames, thus better filenames being checked, but codespell doesn't care about those , at least if I run on `❯ codespell thi.ng`).

**Comment by DimitriPapadopoulos** [comment]
So your suggestion would be to skip URIs, at least **by default**, wouldn't it?

**Comment by yarikoptic** [comment]
yes

**Comment by Borda** [comment]
> So your suggestion would be to skip URIs, at least **by default**, wouldn't it?

yes because URL, in most cases, is not your own creation; fixing them as typo makes them invalid... and you do not want to add any words from strange URL because it will pollute your remeaning checks... :thinking:
